apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -794,15 +805,135 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
     return diagnostic.category;
   }
 
+  private _getProgramWithPatchedEmit<Program extends TTypescript.BuilderProgram | TTypescript.Program>(
+    ts: ExtendedTypeScript,
+    program: Program
+  ): Program {
+    if ((program as Program & { __heftPatched?: boolean }).__heftPatched) {
+      return program;
+    }
+    (program as Program & { __heftPatched?: boolean }).__heftPatched = true;
+
+    const originalEmit: Program['emit'] = program.emit;
+    program.emit = (...args) => {
+      const defaultModuleKind: TTypescript.ModuleKind = this._defaultModuleKind();
+
+      const writeFile: TTypescript.WriteFileCallback | undefined = args[1];
+
+      function wrapWriteFile(
+        jsExtensionOverride: string | undefined
+      ): TTypescript.WriteFileCallback | undefined {
+        if (writeFile === undefined || !jsExtensionOverride) {
+          return writeFile;
+        }
+
+        const replacementExtension: string = `${jsExtensionOverride}$1`;
+        return (
+          fileName: string,
+          data: string,
+          writeBOM: boolean,
+          onError?: ((message: string) => void) | undefined,
+          sourceFiles?: readonly TTypescript.SourceFile[] | undefined
+        ) => {
+          return writeFile(
+            fileName.replace(/\.js(\.map)?$/g, replacementExtension),
+            data,
+            writeBOM,
+            onError,
+            sourceFiles
+          );
+        };
+      }
+
+      const compilerOptions: TTypescript.CompilerOptions = program.getCompilerOptions();
+
+      function runWithSaveRestoreOptions<T>(fn: () => T): T {
```

Comment:
Would it be safer to override `program.getCompilerOptions()` and return a value from a closure? That way we don't have to validate the reversibility of the mutation of the options structure.

Reply:
I'm pretty sure no; we take reference to this internally so you have to hack the object.

@weswigham suggested this when we were going through the different patchers, maybe there's a better way or I misinterpreted his suggestion, but I think this is what we thought might work.

Reply:
It looks like you only bind to `program.getCompilerOptions` in the `getEmitHost` function: https://github.com/microsoft/TypeScript/blob/main/src/compiler/program.ts#L2486
So replacing said function ought to work, as `getEmitHost` is called inside of `emit`

Reply:
Sure, I guess I'll try it like that.

Reply:
Ah, found the binding to `options` in `emitWorker`; it's passed to the getEmitResolver call.

Reply:
Yeah, that's the one I remembered seeing but promptly forgot about.